### PR TITLE
Transformations: Add 'transpose' transform

### PR DESCRIPTION
**Pull request by jmdane**
**What is this feature?**

This feature transposes the data, changes rows to columns, and columns to rows.

**Why do we need this feature?**

Particularly for barcharts this is very useful in cases where you cannot transpose the data to visualize the data correctly. For example:
| env| january| february|
| --- | --- | --- |
| prod | 1 | 2 |
| dev | 3 | 4 |

This will display incorrectly in barcharts unless you do some preprocessing on data you might not be able to. And the desired outcome is: 
| env| prod| dev|
| --- | --- | --- |
| january| 1 | 3 |
| february| 2 | 4 |

**Who is this feature for?**

For users who might not be able to do preprocessing before visualization, AWS Athena users not being able to use PIVOT in queries.

**Which issue(s) does this PR fix?**:

Fixes # None

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: packages/grafana-data/src/transformations/transformers/transpose.test.ts [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        {
+          name: 'env',
+          type: FieldType.string,
+          values: ['january', 'february', 'type'],
+          config: {},
+        },
+        {
+          name: 'dev',
+          type: FieldType.string,
+          values: [11, 6, 'metricA'],
```
Comment (ryantxu) [inline packages/grafana-data/src/transformations/transformers/transpose.test.ts:94]: in this model, all values have to be strings 😢 -- we can not mix numbers and strings in a single array

**Comment by ryantxu** [comment]
Thanks @jmdane -- I have looked at this a few times and then get hung up on the field names and mixing value types in an array  -- In your example, the result is a field with name env and values january,february 🤯  

What behavior should we do when there are no string fields?  EG

| A | B | C | D |
| --- | --- | --- | --- |
| 1.2  | 3.4 | 5.6 | 7.8 |
| 9.0  |  0.1  | 0.2 | 0.3 |

It kinda makes sense to end with:

| Field | Value1 | Value2  |
| --- | --- | --- |
| A  | 1.2 | 9.0 |
| B  |  3.4  | 0.1 |
| C  |  5.6  | 0.2 |
| C  |  7.8  | 0.3 |

And what about:

| time| value |
| --- | --- |
| 2024-06-10 08:30:00 |  5 |
| 2024-06-10 08:31:00 |  7 |
| 2024-06-10 08:32:00 |  7 |

----

Given the typing issues, this has felt more like a visualization setting that a transformation to me -- EG, something that should be supported in barchart and table panels rather than transformation because the type system does not support multiple types across the same column.

**Comment by jmdane** [comment]
I added the option to add new column names and to rename the first column name. 

Functionality is the types are always turned to string if multiple types are present in one column, otherwise, it will use that column type.


time | value
-- | --
2024-06-10 08:30:00 | 5
2024-06-10 08:31:00 | 7
2024-06-10 08:32:00 | 7

Turns into

time | 2024-06-10 08:30:00 | 2024-06-10 08:31:00 | 2024-06-10 08:32:00
-- | -- | -- | --
value |5 | 7 | 7

---

As for visualization setting for certain graph, that would work but I believe it will still implement this same functionality but in the graph.

**Review by imatwawana — COMMENTED** [review]
Thanks for making the changes to the docs! I've made a note about how the markdown file in this PR should be updated as I'm not sure what method was used.

**Comment by jmdane** [comment]
Any updates on this?

**Comment by imatwawana** [comment]
@ryantxu - Are your concerns addressed? 

Also please let me take one last look at it before merging so I can make sure the transform-data.md and associated typescript file are properly updated.

**Review by ryantxu — APPROVED** [review]
This looks good! thanks and sorry for the delay

**Review by baldm0mma — APPROVED** [review]

**Comment by imatwawana** [comment]
@baldm0mma or @ryantxu can you take a look at the build failures on this?

**Comment by baldm0mma** [comment]
> @baldm0mma or @ryantxu can one of you take a look at the failing checks on this?

@jmdane looks like there's a few errors in the drone pipeline. One is in `TransposeTransformerEditor.tsx`, which looks like the `React` import is superfluous. And the second is your test suite seems to be failing, again via an incorrect import:
```
FAIL packages/grafana-data/src/transformations/transformers/transpose.test.ts
  ● Test suite failed to run

    Cannot find module '../../types' from 'packages/grafana-data/src/transformations/transformers/transpose.test.ts'

      1 | import { toDataFrame } from '../../dataframe/processDataFrame';
    > 2 | import { DataTransformerConfig, FieldType } from '../../types';
        | ^
      3 | import { mockTransformationsRegistry } from '../../utils/tests/mockTransformationsRegistry';
      4 | import { transformDataFrame } from '../transformDataFrame';
      5 |

      at Resolver._throwModNotFoundError (node_modules/jest-resolve/build/resolver.js:427:11)
      at Object.<anonymous> (packages/grafana-data/src/transformations/transformers/transpose.test.ts:2:1)
```

**Comment by leeoniya** [comment]
thinking out loud...

> And the desired outcome is: 

| env| prod| dev|
| --- | --- | --- |
| january| 1 | 3 |
| february| 2 | 4 |

i wonder if the version below is more Grafana-like and consistent with similar transformations, where the env is moved into labels and we would allow customizing the field names

| Month | Value {env="prod"} | Value {env="dev"} |
| --- | --- | --- |
| january| 1 | 3 |
| february| 2 | 4 |

e.g. our Partition by values transformation creates frames with the unique values as field labels. it has an additional option to not create labels, as well.

image

**Comment by ryantxu** [comment]
@leeoniya -- I like the label option, but would also need an input for what to call the first field.  Eg "Month" in this example is not in the original dataset

**Comment by leeoniya** [comment]
@jmdane @ryantxu 

pushed a few updates:

1. remove `addNewFields` option and try to handle this automatically (when first field is not time, string, or enum)
2. rename `renameFirstField` -> `firstFieldName` (defaults to "Field")
3. use "Value" field name for non-header fields and use field label to retain both header column name and label value
4. allow renaming of "Value" name for non-header fields.
5. use label `{ row: <rowIndex> }` for Value fields to indicate origin


<details><summary>transpose2.json</summary>

```json
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": {
          "type": "grafana",
          "uid": "-- Grafana --"
        },
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "fiscalYearStartMonth": 0,
  "graphTooltip": 0,
  "id": 322,
  "links": [],
  "panels": [
    {
      "datasource": {
        "default": true,
        "type": "grafana-testdata-datasource",
        "uid": "PD8C576611E62080A"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "thresholds"
          },
          "custom": {
            "align": "auto",
            "cellOptions": {
              "type": "auto"
            },
            "inspect": false
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 8,
        "w": 5,
        "x": 0,
        "y": 0
      },
      "id": 1,
      "options": {
        "cellHeight": "sm",
        "footer": {
          "countRows": false,
          "fields": "",
          "reducer": [
            "sum"
          ],
          "show": false
        },
        "showHeader": true
      },
      "pluginVersion": "11.2.0-pre",
      "targets": [
        {
          "csvContent": "env,january,february\nprod,1,2\ndev,3,4",
          "datasource": {
            "type": "grafana-testdata-datasource",
            "uid": "PD8C576611E62080A"
          },
          "refId": "A",
          "scenarioId": "csv_content"
        }
      ],
      "title": "data 1",
      "type": "table"
    },
    {
      "datasource": {
        "default": false,
        "type": "datasource",
        "uid": "-- Dashboard --"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "thresholds"
          },
          "custom": {
            "align": "auto",
            "cellOptions": {
              "type": "auto"
            },
            "inspect": false
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 8,
        "w": 5,
        "x": 5,
        "y": 0
      },
      "id": 2,
      "options": {
        "cellHeight": "sm",
        "footer": {
          "countRows": false,
          "fields": "",
          "reducer": [
            "sum"
          ],
          "show": false
        },
        "showHeader": true
      },
      "pluginVersion": "11.2.0-pre",
      "targets": [
        {
          "datasource": {
            "type": "datasource",
            "uid": "-- Dashboard --"
          },
          "panelId": 1,
          "refId": "A"
        }
      ],
      "title": "transpose",
      "transformations": [
        {
          "id": "transpose",
          "options": {
            "renameFirstField": "Month"
          }
        }
      ],
      "type": "table"
    },
    {
      "datasource": {
        "default": false,
        "type": "datasource",
        "uid": "-- Dashboard --"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisBorderShow": false,
            "axisCenteredZero": false,
            "axisColorMode": "text",
            "axisLabel": "",
            "axisPlacement": "auto",
            "fillOpacity": 80,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "lineWidth": 1,
            "scaleDistribution": {
              "type": "linear"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 8,
        "w": 5,
        "x": 10,
        "y": 0
      },
      "id": 9,
      "options": {
        "barRadius": 0,
        "barWidth": 0.97,
        "fullHighlight": false,
        "groupWidth": 0.7,
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom",
          "showLegend": true
        },
        "orientation": "auto",
        "showValue": "auto",
        "stacking": "none",
        "tooltip": {
          "mode": "single",
          "sort": "none"
        },
        "xTickLabelRotation": 0,
        "xTickLabelSpacing": 0
      },
      "pluginVersion": "11.2.0-pre",
      "targets": [
        {
          "datasource": {
            "type": "datasource",
            "uid": "-- Dashboard --"
          },
          "panelId": 1,
          "refId": "A"
        }
      ],
      "title": "transpose",
      "transformations": [
        {
          "id": "transpose",
          "options": {
            "renameFirstField": "Month"
          }
        }
      ],
      "type": "barchart"
    },
    {
      "datasource": {
        "default": true,
        "type": "grafana-testdata-datasource",
        "uid": "PD8C576611E62080A"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "thresholds"
          },
          "custom": {
            "align": "auto",
            "cellOptions": {
              "type": "auto"
            },
            "inspect": false
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 8,
        "w": 5,
        "x": 0,
        "y": 8
      },
      "id": 4,
      "options": {
        "cellHeight": "sm",
        "footer": {
          "countRows": false,
          "fields": "",
          "reducer": [
            "sum"
          ],
          "show": false
        },
        "showHeader": true
      },
      "pluginVersion": "11.2.0-pre",
      "targets": [
        {
          "csvContent": "A,B\n1.2,3.4\n9.0,0.1",
          "datasource": {
            "type": "grafana-testdata-datasource",
            "uid": "PD8C576611E62080A"
          },
          "refId": "A",
          "scenarioId": "csv_content"
        }
      ],
      "title": "data 2",
      "type": "table"
    },
    {
      "datasource": {
        "default": false,
        "type": "datasource",
        "uid": "-- Dashboard --"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "thresholds"
          },
          "custom": {
            "align": "auto",
            "cellOptions": {
              "type": "auto"
            },
            "inspect": false
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 8,
        "w": 5,
        "x": 5,
        "y": 8
      },
      "id": 5,
      "options": {
        "cellHeight": "sm",
        "footer": {
          "countRows": false,
          "fields": "",
          "reducer": [
            "sum"
          ],
          "show": false
        },
        "showHeader": true
      },
      "pluginVersion": "11.2.0-pre",
      "targets": [
        {
          "datasource": {
            "type": "datasource",
            "uid": "-- Dashboard --"
          },
          "panelId": 4,
          "refId": "A"
        }
      ],
      "title": "transpose",
      "transformations": [
        {
          "id": "transpose",
          "options": {}
        }
      ],
      "type": "table"
    },
    {
      "datasource": {
        "default": false,
        "type": "datasource",
        "uid": "-- Dashboard --"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisBorderShow": false,
            "axisCenteredZero": false,
            "axisColorMode": "text",
            "axisLabel": "",
            "axisPlacement": "auto",
            "fillOpacity": 80,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "lineWidth": 1,
            "scaleDistribution": {
              "type": "linear"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 8,
        "w": 5,
        "x": 10,
        "y": 8
      },
      "id": 10,
      "options": {
        "barRadius": 0,
        "barWidth": 0.97,
        "fullHighlight": false,
        "groupWidth": 0.7,
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom",
          "showLegend": true
        },
        "orientation": "auto",
        "showValue": "auto",
        "stacking": "none",
        "tooltip": {
          "mode": "single",
          "sort": "none"
        },
        "xTickLabelRotation": 0,
        "xTickLabelSpacing": 0
      },
      "pluginVersion": "11.2.0-pre",
      "targets": [
        {
          "datasource": {
            "type": "datasource",
            "uid": "-- Dashboard --"
          },
          "panelId": 4,
          "refId": "A"
        }
      ],
      "title": "transpose",
      "transformations": [
        {
          "id": "transpose",
          "options": {}
        }
      ],
      "type": "barchart"
    },
    {
      "datasource": {
        "default": true,
        "type": "grafana-testdata-datasource",
        "uid": "PD8C576611E62080A"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "thresholds"
          },
          "custom": {
            "align": "auto",
            "cellOptions": {
              "type": "auto"
            },
            "inspect": false
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 8,
        "w": 5,
        "x": 0,
        "y": 16
      },
      "id": 7,
      "options": {
        "cellHeight": "sm",
        "footer": {
          "countRows": false,
          "fields": "",
          "reducer": [
            "sum"
          ],
          "show": false
        },
        "showHeader": true
      },
      "pluginVersion": "11.2.0-pre",
      "targets": [
        {
          "csvContent": "time,A,B\n2024-06-10 08:30:00,5,4\n2024-06-10 08:31:00,7,5\n2024-06-10 08:32:00,7,6",
          "datasource": {
            "type": "grafana-testdata-datasource",
            "uid": "PD8C576611E62080A"
          },
          "refId": "A",
          "scenarioId": "csv_content"
        }
      ],
      "title": "data 3",
      "type": "table"
    },
    {
      "datasource": {
        "default": false,
        "type": "datasource",
        "uid": "-- Dashboard --"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "thresholds"
          },
          "custom": {
            "align": "auto",
            "cellOptions": {
              "type": "auto"
            },
            "inspect": false
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 8,
        "w": 5,
        "x": 5,
        "y": 16
      },
      "id": 8,
      "options": {
        "cellHeight": "sm",
        "footer": {
          "countRows": false,
          "fields": "",
          "reducer": [
            "sum"
          ],
          "show": false
        },
        "showHeader": true
      },
      "pluginVersion": "11.2.0-pre",
      "targets": [
        {
          "datasource": {
            "type": "datasource",
            "uid": "-- Dashboard --"
          },
          "panelId": 7,
          "refId": "A"
        }
      ],
      "title": "transpose",
      "transformations": [
        {
          "id": "transpose",
          "options": {}
        }
      ],
      "type": "table"
    },
    {
      "datasource": {
        "default": false,
        "type": "datasource",
        "uid": "-- Dashboard --"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisBorderShow": false,
            "axisCenteredZero": false,
            "axisColorMode": "text",
            "axisLabel": "",
            "axisPlacement": "auto",
            "fillOpacity": 80,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "lineWidth": 1,
            "scaleDistribution": {
              "type": "linear"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 8,
        "w": 5,
        "x": 10,
        "y": 16
      },
      "id": 11,
      "options": {
        "barRadius": 0,
        "barWidth": 0.97,
        "fullHighlight": false,
        "groupWidth": 0.7,
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom",
          "showLegend": true
        },
        "orientation": "auto",
        "showValue": "auto",
        "stacking": "none",
        "tooltip": {
          "mode": "single",
          "sort": "none"
        },
        "xTickLabelRotation": 0,
        "xTickLabelSpacing": 0
      },
      "pluginVersion": "11.2.0-pre",
      "targets": [
        {
          "datasource": {
            "type": "datasource",
            "uid": "-- Dashboard --"
          },
          "panelId": 7,
          "refId": "A"
        }
      ],
      "title": "transpose",
      "transformations": [
        {
          "id": "transpose",
          "options": {
            "renameFirstField": "Server"
          }
        }
      ],
      "type": "barchart"
    }
  ],
  "schemaVersion": 39,
  "tags": [],
  "templating": {
    "list": []
  },
  "time": {
    "from": "now-6h",
    "to": "now"
  },
  "timepicker": {},
  "timezone": "browser",
  "title": "transpose2",
  "uid": "fducg0t0ztx4wd",
  "version": 5,
  "weekStart": ""
}
```
</details>

before:

image


after:

image

**Comment by diegommm** [comment]
Hi! Greetings from the Release Guild. This is a friendly reminder that the Steady [11.2.x Release Branch](https://github.com/grafana/grafana/tree/v11.2.x) was cut this week, and that it includes the changes committed up to https://github.com/grafana/grafana/commit/b4126d3bce3627e9460e2db0e870c8d984be8e84. This PR was merged after it, so we wanted to let you know that you can add the `backport v11.2.x` label to to backport it and make it part of the next 11.2.x Patch Release, just in case you may have missed the news.

Thank you!

**Comment by slhck** [comment]
Did this land in 11.2.1 now? I didn't see it in the changelog.

**Comment by slhck** [comment]
When will this feature be available?

**Comment by HorselessName** [comment]
I am having a problem and this implementation would resolve it.

https://community.grafana.com/t/how-can-i-join-two-rows-into-one-rows-move-columns-from-row-b-to-row-a/139175

It's the exact solution for my exact problem. :c 
Will it ever be implemented?

**Pull request merged**